src/GeckofxHtmlToPDFComponent.cs

```diff
@@ -446,6 +462,158 @@ public void OnSecurityChange(nsIWebProgress aWebProgress, nsIRequest aRequest, u
+
+				gsProcess.WaitForExit(600000);		// timeout - 10 minutes? 1 minutes? 5 minutes? 30 sec?
+				gsProcess.OutputDataReceived -= ShrinkOutputHandler;
+
+				if (File.Exists(newTempFile) && gsProcess.ExitCode == 0)
+				{
+					var oldInfo = new FileInfo(_pathToTempPdf);
+					var newInfo = new FileInfo(newTempFile);
+					// If the process actually made the file larger, ignore the result (and delete it).
+					if (newInfo.Length < oldInfo.Length)
+						return newTempFile;
+					File.Delete(newTempFile);
+				}
+			}
+			catch (Exception ex)
+			{
+				// If something bad happens, well, that means we can't shrink the file.  Keep going.
+				// On Windows, we should tell the user about downloading ghostscript.
+				// Except that this gets way too intrusive quickly.  We need to tell the user at least
+				// once, but how do we control (not) repeating the message?
+				//if (info.FileName == "gswin32.exe")
+				//{
+				//	var msg = "Cannot shrink the PDF because ghostscript cannot be found on this computer.  It can be downloaded from https://ghostscript.com/download/.";
+				//	var header = "Cannot shrink PDF";
+				//	MessageBox.Show(msg, header, MessageBoxButtons.OK, MessageBoxIcon.Information);
+				//}
+			}
+			return null;
+		}
+
+		/// <summary>
+		/// Try to find ghostscript on the Windows system, looking in the likeliest installation
+		/// locations and trying to not depend on the exact version installed.
+		/// </summary>
+		/// <remarks>
+		/// Ghostscript 9.21 for 64-bit Windows 10 installs by default as
+		/// C:\Program Files\gs\gs9.21\bin\gswin64c.exe or C:\Program Files\gs\gs9.21\bin\gswin64.exe.
+		/// The former uses the current console window, the latter pops up its own command window.
+		/// </remarks>
+		private string FindGhostcriptOnWindows()
+		{
+			var baseName = "gswin32";
+			if (Environment.Is64BitOperatingSystem)
+				baseName = "gswin64";
+			// Look for the four most common installation locations.
+			var baseDir = @"C:\Program Files\gs";
+			if (!Directory.Exists(baseDir))
+				baseDir = @"C:\Program Files (x86)\gs";
+			if (!Directory.Exists(baseDir))
+				baseDir = @"D:\Program Files\gs";
+			if (!Directory.Exists(baseDir))
+				baseDir = @"D:\Program Files (x86)\gs";
+			if (!Directory.Exists(baseDir))
+				return baseName + ".exe";   // almost certainly won't work, but we can't scan the whole disk!
+			foreach (var versionDir in Directory.GetDirectories(baseDir))
+			{
+				var prog = Path.Combine(versionDir, "bin", baseName + "c.exe");
+				if (File.Exists(prog))
+					return prog;
+				prog = Path.Combine(versionDir, "bin", baseName + ".exe");
+				if (File.Exists(prog))
+					return prog;
+			}
+			return baseName + ".exe";
+		}
+
+		int _firstPage;
+		int _numPages;
+
+		/// <summary>
+		/// Parse the standard output from ghostscript to feed the progress reporting.
+		/// </summary>
+		private void ShrinkOutputHandler(object sendingProcess, DataReceivedEventArgs outputLine)
```

Comment:
How long does this take? If it's not, like, instant, then this feature would be better as a separate process, probably called from Bloom. I say that because we could just not do it until you ask Bloom to Save the PDF.  Same goes for CMYK conversion.